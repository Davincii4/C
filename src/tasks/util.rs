pub(super) mod advisory_lock;
