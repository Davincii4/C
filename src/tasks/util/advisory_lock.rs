use std::error::Error;

use diesel::prelude::*;
use diesel::sql_types::BigInt;
use diesel::PgConnection;

sql_function!(fn pg_try_advisory_lock(key: BigInt) -> Bool);
sql_function!(fn pg_advisory_unlock(key: BigInt) -> Bool);

/// Run the callback if the session advisory lock for the given key can be obtained.
///
/// If the lock is not already held, the callback will be called and the lock will be unlocked
/// after the closure returns.  If the lock is already held, the function returns an error without
/// calling the callback.
pub(crate) fn with_advisory_lock<F>(
    conn: &PgConnection,
    key: i64,
    f: F,
) -> Result<(), Box<dyn Error>>
where
    F: FnOnce(&PgConnection) -> QueryResult<()>,
{
    if !diesel::select(pg_try_advisory_lock(key)).get_result(conn)? {
        let string = format!(
            "A job holding the session advisory lock for key {} is already running",
            key
        );
        println!("return");
        return Err(string.into());
    }
    println!("umm");
    let _dont_drop_yet = DropGuard { conn, key };
    f(conn).map_err(Into::into)
}

struct DropGuard<'a> {
    conn: &'a PgConnection,
    key: i64,
}

impl<'a> Drop for DropGuard<'a> {
    fn drop(&mut self) {
        match diesel::select(pg_advisory_unlock(self.key)).get_result(self.conn) {
            Ok(true) => (),
            Ok(false) => println!(
                "Error: job advisory lock for key {} was not locked",
                self.key
            ),
            Err(err) => println!("Error unlocking advisory lock (key: {}): {}", self.key, err),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_util::*;

    #[test]
    fn lock_released_after_callback_returns() {
        const KEY: i64 = -1;
        let conn1 = pg_connection();
        let conn2 = pg_connection();

        let mut callback_run = false;
        let result = with_advisory_lock(&conn1, KEY, |_| {
            // Another connection cannot obtain the lock
            assert!(!diesel::select(pg_try_advisory_lock(KEY)).get_result(&conn2)?);
            callback_run = true;
            Ok(())
        });
        assert!(result.is_ok());
        assert!(callback_run);

        // Another connection can now obtain the lock
        assert_eq!(
            diesel::select(pg_try_advisory_lock(KEY)).get_result(&conn2),
            Ok(true)
        );
    }

    #[test]
    fn test_already_locked() {
        const KEY: i64 = -2;
        let conn1 = pg_connection();
        let conn2 = pg_connection();

        // Another connection obtains the lock first
        assert_eq!(
            diesel::select(pg_try_advisory_lock(KEY)).get_result(&conn2),
            Ok(true)
        );

        let mut callback_run = false;
        let result = with_advisory_lock(&conn1, KEY, |_| {
            callback_run = true;
            Ok(())
        });
        assert!(dbg!(result).is_err());
        assert!(!callback_run);
    }
}
