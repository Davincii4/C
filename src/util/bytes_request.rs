use crate::middleware::log_request::ErrorField;
use axum::body::Bytes;
use axum::extract::FromRequest;
use axum::response::{IntoResponse, Response};
use axum::{async_trait, Extension, RequestExt};
use http::{Request, StatusCode};
use http_body::{Body, LengthLimitError};
use std::error::Error;
use std::ops::{Deref, DerefMut};

#[derive(Debug)]
pub struct BytesRequest(pub Request<Bytes>);

impl Deref for BytesRequest {
    type Target = Request<Bytes>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for BytesRequest {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

#[async_trait]
impl<S, B> FromRequest<S, B> for BytesRequest
where
    S: Send + Sync,
    B: Body + Send + 'static,
    B::Data: Send,
    B::Error: Into<Box<dyn Error + Send + Sync>>,
{
    type Rejection = Response;

    async fn from_request(req: Request<B>, _state: &S) -> Result<Self, Self::Rejection> {
        let request = match req.with_limited_body() {
            Ok(req) => {
                let (parts, body) = req.into_parts();

                let bytes = hyper::body::to_bytes(body).await.map_err(|err| {
                    if err.downcast_ref::<LengthLimitError>().is_some() {
                        StatusCode::BAD_REQUEST.into_response()
                    } else {
                        server_error_response(&*err)
                    }
                })?;

                Request::from_parts(parts, bytes)
            }
            Err(req) => {
                let (parts, body) = req.into_parts();

                let bytes = hyper::body::to_bytes(body)
                    .await
                    .map_err(|err| server_error_response(&*err.into()))?;

                Request::from_parts(parts, bytes)
            }
        };

        Ok(BytesRequest(request))
    }
}

/// Logs an error message and returns a generic status 500 response
fn server_error_response<E: Error + ?Sized>(error: &E) -> Response {
    error!(%error, "Internal Server Error");

    sentry::capture_error(error);

    (
        StatusCode::INTERNAL_SERVER_ERROR,
        Extension(ErrorField(error.to_string())),
        "Internal Server Error",
    )
        .into_response()
}

#[cfg(test)]
mod tests {
    use super::BytesRequest;
    use axum::extract::DefaultBodyLimit;
    use axum::routing::get;
    use axum::Router;
    use http::{Request, StatusCode};
    use tower::ServiceExt;

    #[tokio::test]
    async fn content_length_too_large() {
        const BODY_SIZE_LIMIT: usize = 4096;

        fn app() -> Router {
            async fn bytes_request(_req: BytesRequest) {}

            Router::new()
                .route("/", get(bytes_request))
                .layer(DefaultBodyLimit::max(BODY_SIZE_LIMIT))
        }

        let body = vec![0; BODY_SIZE_LIMIT + 1];
        let body = axum::body::Body::from(body);
        let request = Request::get("/").body(body).unwrap();
        let response = app().oneshot(request).await.unwrap();

        assert_eq!(response.status(), StatusCode::BAD_REQUEST);

        let body = vec![0; BODY_SIZE_LIMIT];
        let body = axum::body::Body::from(body);
        let request = Request::get("/").body(body).unwrap();
        let response = app().oneshot(request).await.unwrap();

        assert_eq!(response.status(), StatusCode::OK);
    }
}
