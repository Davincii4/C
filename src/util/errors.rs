//! This module implements several error types and traits.
//!
//! The suggested usage in returned results is as follows:
//!
//! * The concrete `util::concrete::Error` type (re-exported as `util::Error`) is great for code
//!   that is not part of the request/response lifecycle.  It avoids pulling in the unnecessary
//!   infrastructure to convert errors into a user facing JSON responses (relative to `AppError`).
//! * `diesel::QueryResult` - There is a lot of code that only deals with query errors.  If only
//!   one type of error is possible in a function, using that specific error is preferable to the
//!   more general `util::Error`.  This is especially common in model code.
//! * `util::errors::AppResult` - Some failures should be converted into user facing JSON
//!   responses.  This error type is more dynamic and is box allocated.  Low-level errors are
//!   typically not converted to user facing errors and most usage is within the models,
//!   controllers, and middleware layers.

use axum::response::IntoResponse;
use std::any::{Any, TypeId};
use std::error::Error;
use std::fmt;

use axum::Extension;
use chrono::NaiveDateTime;
use diesel::result::Error as DieselError;
use http::StatusCode;
use tokio::task::JoinError;

use crate::db::PoolError;
use crate::middleware::log_request::{CauseField, ErrorField};

mod json;

pub use json::TOKEN_FORMAT_ERROR;
pub(crate) use json::{
    InsecurelyGeneratedTokenRevoked, MetricsDisabled, NotFound, OwnershipInvitationExpired,
    ReadOnlyMode, RouteBlocked, TooManyRequests,
};

pub type BoxedAppError = Box<dyn AppError>;

/// Returns an error with status 200 and the provided description as JSON
///
/// This is for backwards compatibility with cargo endpoints.  For all other
/// endpoints, use helpers like `bad_request` or `server_error` which set a
/// correct status code.
pub fn cargo_err<S: ToString>(error: S) -> BoxedAppError {
    Box::new(json::Ok(error.to_string()))
}

// The following are intended to be used for errors being sent back to the Ember
// frontend, not to cargo as cargo does not handle non-200 response codes well
// (see <https://github.com/rust-lang/cargo/issues/3995>), but Ember requires
// non-200 response codes for its stores to work properly.

/// Return an error with status 400 and the provided description as JSON
pub fn bad_request<S: ToString>(error: S) -> BoxedAppError {
    Box::new(json::BadRequest(error.to_string()))
}

pub fn account_locked(reason: &str, until: Option<NaiveDateTime>) -> BoxedAppError {
    Box::new(json::AccountLocked {
        reason: reason.to_string(),
        until,
    })
}

pub fn forbidden() -> BoxedAppError {
    Box::new(json::Forbidden)
}

pub fn not_found() -> BoxedAppError {
    Box::new(json::NotFound)
}

/// Returns an error with status 500 and the provided description as JSON
pub fn server_error<S: ToString>(error: S) -> BoxedAppError {
    Box::new(json::ServerError(error.to_string()))
}

/// Returns an error with status 503 and the provided description as JSON
pub fn service_unavailable() -> BoxedAppError {
    Box::new(json::ServiceUnavailable)
}

// =============================================================================
// AppError trait

pub trait AppError: Send + fmt::Display + fmt::Debug + 'static {
    /// Generate an HTTP response for the error
    ///
    /// If none is returned, the error will bubble up the middleware stack
    /// where it is eventually logged and turned into a status 500 response.
    fn response(&self) -> axum::response::Response;

    /// The cause of an error response
    ///
    /// If present, an error provided to the `LogRequests` middleware.
    ///
    /// This is intended for use with the `ChainError` trait, where a user facing
    /// error may wrap an internal error we still want to log.
    fn cause(&self) -> Option<&dyn AppError> {
        None
    }

    fn get_type_id(&self) -> TypeId {
        TypeId::of::<Self>()
    }

    fn chain<E>(self, error: E) -> BoxedAppError
    where
        Self: Sized,
        E: AppError,
    {
        Box::new(ChainedError {
            error,
            cause: Box::new(self),
        })
    }
}

impl dyn AppError {
    pub fn is<T: Any>(&self) -> bool {
        self.get_type_id() == TypeId::of::<T>()
    }
}

impl AppError for BoxedAppError {
    fn response(&self) -> axum::response::Response {
        (**self).response()
    }

    fn cause(&self) -> Option<&dyn AppError> {
        (**self).cause()
    }

    fn get_type_id(&self) -> TypeId {
        (**self).get_type_id()
    }
}

impl IntoResponse for BoxedAppError {
    fn into_response(self) -> axum::response::Response {
        let mut response = self.response();

        if let Some(cause) = self.cause() {
            response
                .extensions_mut()
                .insert(CauseField(cause.to_string()));
        }

        response
    }
}

pub type AppResult<T> = Result<T, BoxedAppError>;

// =============================================================================
// Chaining errors

#[derive(Debug)]
struct ChainedError<E> {
    error: E,
    cause: BoxedAppError,
}

impl<E: AppError> AppError for ChainedError<E> {
    fn response(&self) -> axum::response::Response {
        self.error.response()
    }

    fn cause(&self) -> Option<&dyn AppError> {
        Some(&*self.cause)
    }
}

impl<E: AppError> fmt::Display for ChainedError<E> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} caused by {}", self.error, self.cause)
    }
}

// =============================================================================
// Error impls

impl<E: Error + Send + 'static> AppError for E {
    fn response(&self) -> axum::response::Response {
        error!(error = %self, "Internal Server Error");

        sentry::capture_error(self);

        server_error_response(self.to_string())
    }
}

impl From<base64::DecodeError> for BoxedAppError {
    fn from(err: base64::DecodeError) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<diesel::ConnectionError> for BoxedAppError {
    fn from(err: diesel::ConnectionError) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<DieselError> for BoxedAppError {
    fn from(err: DieselError) -> BoxedAppError {
        match err {
            DieselError::NotFound => not_found(),
            DieselError::DatabaseError(_, info)
                if info.message().ends_with("read-only transaction") =>
            {
                Box::new(ReadOnlyMode)
            }
            _ => Box::new(err),
        }
    }
}

impl From<http::Error> for BoxedAppError {
    fn from(err: http::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<lettre::error::Error> for BoxedAppError {
    fn from(err: lettre::error::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<lettre::address::AddressError> for BoxedAppError {
    fn from(err: lettre::address::AddressError) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<PoolError> for BoxedAppError {
    fn from(err: PoolError) -> BoxedAppError {
        match err {
            PoolError::UnhealthyPool => service_unavailable(),
            _ => Box::new(err),
        }
    }
}

impl From<prometheus::Error> for BoxedAppError {
    fn from(err: prometheus::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<reqwest::Error> for BoxedAppError {
    fn from(err: reqwest::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<serde_json::Error> for BoxedAppError {
    fn from(err: serde_json::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<std::io::Error> for BoxedAppError {
    fn from(err: std::io::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<crates_io_worker::EnqueueError> for BoxedAppError {
    fn from(err: crates_io_worker::EnqueueError) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<JoinError> for BoxedAppError {
    fn from(err: JoinError) -> BoxedAppError {
        Box::new(err)
    }
}

// =============================================================================
// Internal error for use with `chain_error`

#[derive(Debug)]
struct InternalAppError {
    description: String,
}

impl fmt::Display for InternalAppError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.description)?;
        Ok(())
    }
}

impl AppError for InternalAppError {
    fn response(&self) -> axum::response::Response {
        error!(error = %self.description, "Internal Server Error");

        sentry::capture_message(&self.description, sentry::Level::Error);

        server_error_response(self.description.to_string())
    }
}

#[derive(Debug)]
struct InternalAppErrorStatic {
    description: &'static str,
}

impl fmt::Display for InternalAppErrorStatic {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.description)?;
        Ok(())
    }
}

impl AppError for InternalAppErrorStatic {
    fn response(&self) -> axum::response::Response {
        error!(error = %self.description, "Internal Server Error");

        sentry::capture_message(self.description, sentry::Level::Error);

        server_error_response(self.description.to_string())
    }
}

pub fn internal<S: ToString>(error: S) -> BoxedAppError {
    Box::new(InternalAppError {
        description: error.to_string(),
    })
}

fn server_error_response(error: String) -> axum::response::Response {
    (
        StatusCode::INTERNAL_SERVER_ERROR,
        Extension(ErrorField(error)),
        "Internal Server Error",
    )
        .into_response()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::middleware::log_request::CauseField;
    use axum::response::IntoResponse;
    use diesel::result::Error as DieselError;
    use http::StatusCode;

    #[test]
    fn http_error_responses() {
        use crate::serde::de::Error;

        // Types for handling common error status codes
        assert_eq!(bad_request("").response().status(), StatusCode::BAD_REQUEST);
        assert_eq!(forbidden().response().status(), StatusCode::FORBIDDEN);
        assert_eq!(
            BoxedAppError::from(DieselError::NotFound)
                .response()
                .status(),
            StatusCode::NOT_FOUND
        );
        assert_eq!(not_found().response().status(), StatusCode::NOT_FOUND);

        // cargo_err errors are returned as 200 so that cargo displays this nicely on the command line
        assert_eq!(cargo_err("").response().status(), StatusCode::OK);

        // Inner errors are captured for logging when wrapped by a user facing error
        let response = "-1"
            .parse::<u8>()
            .map_err(|err| err.chain(internal("middle error")))
            .map_err(|err| err.chain(bad_request("outer user facing error")))
            .unwrap_err()
            .into_response();
        assert_eq!(response.status(), StatusCode::BAD_REQUEST);
        assert_eq!(
            response.extensions().get::<CauseField>().unwrap().0,
            "middle error caused by invalid digit found in string"
        );

        // All other error types are converted to internal server errors
        assert_eq!(
            internal("").response().status(),
            StatusCode::INTERNAL_SERVER_ERROR
        );
        assert_eq!(
            BoxedAppError::from(serde_json::Error::custom("ExpectedColon"))
                .response()
                .status(),
            StatusCode::INTERNAL_SERVER_ERROR
        );
        assert_eq!(
            BoxedAppError::from(::std::io::Error::new(::std::io::ErrorKind::Other, ""))
                .response()
                .status(),
            StatusCode::INTERNAL_SERVER_ERROR
        );
    }

    #[test]
    fn chain_error_internal() {
        assert_eq!(
            Err::<(), _>(internal("inner"))
                .map_err(|err| err.chain(internal("middle")))
                .map_err(|err| err.chain(internal("outer")))
                .unwrap_err()
                .to_string(),
            "outer caused by middle caused by inner"
        );
        assert_eq!(
            Err::<(), _>(internal("inner"))
                .map_err(|err| err.chain(internal("outer")))
                .unwrap_err()
                .to_string(),
            "outer caused by inner"
        );

        // Don't do this, the user will see a generic 500 error instead of the intended message
        assert_eq!(
            Err::<(), _>(cargo_err("inner"))
                .map_err(|err| err.chain(internal("outer")))
                .unwrap_err()
                .to_string(),
            "outer caused by inner"
        );
        assert_eq!(
            Err::<(), _>(forbidden())
                .map_err(|err| err.chain(internal("outer")))
                .unwrap_err()
                .to_string(),
            "outer caused by must be logged in to perform that action"
        );
    }

    #[test]
    fn chain_error_user_facing() {
        // Do this rarely, the user will only see the outer error
        assert_eq!(
            Err::<(), _>(cargo_err("inner"))
                .map_err(|err| err.chain(cargo_err("outer")))
                .unwrap_err()
                .to_string(),
            "outer caused by inner" // never logged
        );

        // The outer error is sent as a response to the client.
        // The inner error never bubbles up to the logging middleware
        assert_eq!(
            Err::<(), _>(std::io::Error::from(std::io::ErrorKind::PermissionDenied))
                .map_err(|err| err.chain(cargo_err("outer")))
                .unwrap_err()
                .to_string(),
            "outer caused by permission denied" // never logged
        );
    }
}
