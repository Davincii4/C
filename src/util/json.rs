use std::collections::HashMap;
use std::io::Cursor;

use serde_json;
use serde::Serialize;

use conduit::Response;

pub fn json_response<T: Serialize>(t: &T) -> Response {
    let json = serde_json::to_string(t).unwrap();
    let mut headers = HashMap::new();
    headers.insert(
        "Content-Type".to_string(),
        vec!["application/json; charset=utf-8".to_string()],
    );
    headers.insert("Content-Length".to_string(), vec![json.len().to_string()]);
    Response {
        status: (200, "OK"),
        headers,
        body: Box::new(Cursor::new(json.into_bytes())),
    }
}

#[derive(Serialize)]
struct StringError<'a> {
    detail: &'a str,
}

#[derive(Serialize)]
struct Bad<'a> {
    errors: Vec<StringError<'a>>,
}

pub fn json_error(status: (u32, &'static str), detail: &str) -> Response {
    let mut response = json_error_200(detail);
    response.status = status;
    response
}

pub fn json_error_200(detail: &str) -> Response {
    json_response(&Bad {
        errors: vec![StringError { detail }],
    })
}
