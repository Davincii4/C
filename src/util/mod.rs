pub use self::errors::{bad_request, human, internal, internal_error, CargoError, CargoResult};
pub use self::errors::{std_error, ChainError};
pub use self::io_util::{read_fill, LimitErrorReader, read_le_u32};
pub use self::json::{json_error, json_response, json_error_200};
pub use self::request_proxy::RequestProxy;

pub mod errors;
pub mod rfc3339;
mod io_util;
mod json;
mod request_proxy;
