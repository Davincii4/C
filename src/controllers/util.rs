use super::prelude::*;
use crate::util::errors::{forbidden, internal, AppError, AppResult};
use conduit_router::RequestParams;

/// The Origin header (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin)
/// is sent with CORS requests and POST requests, and indicates where the request comes from.
/// We don't want to accept authenticated requests that originated from other sites, so this
/// function returns an error if the Origin header doesn't match what we expect "this site" to
/// be: https://crates.io in production, or http://localhost:port/ in development.
pub fn verify_origin(req: &dyn RequestExt) -> AppResult<()> {
    let headers = req.headers();
    let allowed_origins = &req.app().config.allowed_origins;

    let bad_origin = headers
        .get_all(header::ORIGIN)
        .iter()
        .find(|value| !allowed_origins.contains(value));

    if let Some(bad_origin) = bad_origin {
        let error_message =
            format!("only same-origin requests can be authenticated. got {bad_origin:?}");
        return Err(internal(&error_message).chain(forbidden()));
    }
    Ok(())
}

pub trait RequestParamExt<'a> {
    fn param(self, key: &str) -> Option<&'a str>;
}

impl<'a> RequestParamExt<'a> for &'a (dyn RequestExt + 'a) {
    fn param(self, key: &str) -> Option<&'a str> {
        self.params().find(key)
    }
}
