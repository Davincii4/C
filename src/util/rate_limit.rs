use conduit::Request;
use std::collections::HashMap;

use crate::db::{DieselPool};
use crate::middleware::app::RequestApp;
use crate::models::{User};
use crate::util::errors::{CargoResult, TooManyRequests};
use std::sync::{Arc, Mutex};

use chrono::{DateTime, Duration, Utc};

/// Settings for a rate-limited route.
#[derive(Debug, Clone)]
pub struct RateLimitSettings {
    /// The code for this category. Can be stored in a database, etc.
    pub key: String,
    /// The maximum number of tokens that can be acquired.
    pub max_amount: usize,
    /// How often we refill
    pub refill_time: Duration,
    /// The number of tokens that are added during a refill.
    pub refill_amount: usize,
}

/// The result from a rate limit check.
#[derive(Debug, Clone, Copy)]
pub struct RateLimitResult {
    /// The remaining number of requests available
    remaining: usize,
}

/// A type that can perform rate limiting.
pub trait RateLimiter {
    fn check_limit_multiple(&self, tokens: u32, user: &User, category: RateLimitCategory) -> CargoResult<RateLimitResult>;
}

/// Rate limit using a postgresql database.
#[allow(missing_debug_implementations)]
#[derive(Clone)]
pub struct RateLimiterPostgres {
    diesel_database: DieselPool,
}

impl RateLimiterPostgres {
    /// Create a new postgres rate limiter from the given database pool.
    pub fn new(diesel_database: DieselPool) -> RateLimiterPostgres {
        RateLimiterPostgres {
            diesel_database,
        }
    }
}

impl RateLimiter for RateLimiterPostgres {
    fn check_limit_multiple(&self, _tokens: u32, _user: &User, _category: RateLimitCategory) -> CargoResult<RateLimitResult> {
        let _conn = self.diesel_database.get()?;

        // TODO: Database interaction.
        Err(Box::new(TooManyRequests))
    }
}

type UserId = i32;
type RateLimiterMemoryKey = (UserId, RateLimitCategory);
#[derive(Debug, Clone, Copy)]
struct RateLimiterMemoryValue {
    pub value: usize,
    // TODO: Time.
    pub last_update: DateTime<Utc>,
}

/// Rate limit using an internal memory store. This may not be ideal in a load-balanced
/// environment, unless all requests from a user get routed to the same instance.
#[derive(Debug, Clone)]
pub struct RateLimiterMemory {
    data: Arc<Mutex<HashMap<RateLimiterMemoryKey, RateLimiterMemoryValue>>>,
}

impl RateLimiterMemory {
    /// Create a new memory-based rate limiter
    pub fn new() -> RateLimiterMemory {
        RateLimiterMemory {
            data: Arc::new(Mutex::new(HashMap::new())),
        }
    }
}

impl RateLimiter for RateLimiterMemory {
    fn check_limit_multiple(&self, tokens: u32, user: &User, category: RateLimitCategory) -> CargoResult<RateLimitResult> {
        let mut data = self.data.lock().unwrap();
        let settings = category.settings();
        let now = Utc::now();
        let mut entry = data.entry((user.id, category)).or_insert_with(|| RateLimiterMemoryValue { value: settings.max_amount, last_update: now });
        println!("Previous entry: {:?}", entry);
        let mut now2 = now;
        let mut refill_count = 0;
        while now2 > entry.last_update + settings.refill_time {
            now2 = now2 - settings.refill_time;
            refill_count += 1;
        }
        entry.value = std::cmp::min(
            settings.max_amount,
            entry.value + refill_count * settings.refill_amount);

        entry.last_update = std::cmp::min(
            now,
            entry.last_update + (settings.refill_time * refill_count as i32));

        if entry.value < tokens as usize {
            return Err(Box::new(TooManyRequests));
        }

        entry.value -= tokens as usize;

        Ok(RateLimitResult { remaining: entry.value })
    }
}

/// A rate limiter that does not limit at all.
#[derive(Debug, Clone, Copy)]
pub struct RateLimiterUnlimited;

impl RateLimiter for RateLimiterUnlimited {
    fn check_limit_multiple(&self, _tokens: u32, _user: &User, category: RateLimitCategory) -> CargoResult<RateLimitResult> {
        println!("Unlimited rate limiter!");
        let settings = category.settings();
        Ok(RateLimitResult { remaining: settings.max_amount })
    }
}

/// All of the possible rate limit buckets. When rate limiting a new endpoint, add it here and set
/// the settings below.
#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
pub enum RateLimitCategory {
    PublishCrate,
    // How often a new crate can be uploaded
    NewCrate,
    // How often an uploaded crate can be uploaded
    NewVersion,
    // How often a request for crate info can be made
    CrateInfo,
}

impl RateLimitCategory {
    pub fn settings(&self) -> RateLimitSettings {
        use RateLimitCategory::*;
        match *self {
            PublishCrate => RateLimitSettings { key: "publish-crate".into(), max_amount: 3, refill_time: Duration::seconds(60), refill_amount: 1 },
            NewCrate => RateLimitSettings { key: "new-crate".into(), max_amount: 3, refill_time: Duration::seconds(60), refill_amount: 1 },
            NewVersion => RateLimitSettings { key: "new_version".into(), max_amount: 60, refill_time: Duration::seconds(1), refill_amount: 1 },
            CrateInfo => RateLimitSettings { key: "crate-info".into(), max_amount: 5, refill_time: Duration::seconds(10), refill_amount: 1 },
        }
    }
}

/// A trait that makes it possible to call `check_rate_limit` directly on a request object.
pub trait RequestRateLimit {
    /// Check the rate limit for the given endpoint. This function consumes a single token from the
    /// token bucket.
    fn check_rate_limit(&mut self, user: &User, category: RateLimitCategory) -> CargoResult<RateLimitResult>;
}

impl<T: Request + ?Sized> RequestRateLimit for T {
    fn check_rate_limit(&mut self, user: &User, category: RateLimitCategory) -> CargoResult<RateLimitResult> {
        let limiter = &self.app().rate_limiter;
        limiter.check_limit_multiple(1, user, category)
    }
}
