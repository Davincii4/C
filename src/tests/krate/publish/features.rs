use crate::builders::{CrateBuilder, DependencyBuilder, PublishBuilder};
use crate::util::{RequestHelper, TestApp};
use http::StatusCode;
use insta::assert_json_snapshot;

#[test]
fn features_version_2() {
    let (app, _, user, token) = TestApp::full().with_token();

    app.db(|conn| {
        // Insert a crate directly into the database so that foo_new can depend on it
        CrateBuilder::new("bar", user.as_model().id).expect_build(conn);
    });

    let dependency = DependencyBuilder::new("bar");

    let crate_to_publish = PublishBuilder::new("foo", "1.0.0")
        .dependency(dependency)
        .feature("new_feat", &["dep:bar", "bar?/feat"])
        .feature("old_feat", &[]);
    token.publish_crate(crate_to_publish).good();

    let crates = app.crates_from_index_head("foo");
    assert_json_snapshot!(crates);
}

#[test]
fn invalid_feature_name() {
    let (app, _, _, token) = TestApp::full().with_token();

    let crate_to_publish = PublishBuilder::new("foo", "1.0.0").feature("~foo", &[]);
    let response = token.publish_crate(crate_to_publish);
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.into_json());
    assert!(app.stored_files().is_empty());
}

#[test]
fn invalid_feature() {
    let (app, _, _, token) = TestApp::full().with_token();

    let crate_to_publish = PublishBuilder::new("foo", "1.0.0").feature("foo", &["!bar"]);
    let response = token.publish_crate(crate_to_publish);
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.into_json());
    assert!(app.stored_files().is_empty());
}
