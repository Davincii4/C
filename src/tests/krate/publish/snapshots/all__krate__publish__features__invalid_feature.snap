---
source: src/tests/krate/publish/features.rs
expression: response.into_json()
---
{
  "errors": [
    {
      "detail": "\"!bar\" is an invalid feature name"
    }
  ]
}
